### ISSUE TITLE
auparse_next_event returns 0 after auparse_feed

### ISSUE THREAD

**Issue by neel** [linux-audit/audit-userspace#416]
Given au initialized with `auparse_init(AUSOURCE_FEED, 0)` the function `auparse_next_event` returns 0 even after passing data to  `auparse_feed` as shown in the example below. This happens even when there is no callback set.

```cpp
if (auparse_feed(_au, l.data(), l.size()) != 0) {
    std::cerr << "Failed to feed data to Auparse. Error: " << std::strerror(errno) << std::endl;
}
auparse_flush_feed(_au);
int res = auparse_next_event(_au); // res is 0
while (res > 0) {
    // never comes here
}
```

This happens because `consume_feed` loops over all the events unconditionally. 

```c
static void consume_feed(auparse_state_t *au, int flush)
{
	//if (debug) printf("consume feed, flush %d\n", flush);
	while (auparse_next_event(au) > 0) {
		if (au->callback) {
			(*au->callback)(au, AUPARSE_CB_EVENT_READY, au->callback_user_data);
		}
	}
     ...
}
```

I'd prefer a non-callback way of extracting events which is achievable by simple change

```c
static void consume_feed(auparse_state_t *au, int flush)
{
	//if (debug) printf("consume feed, flush %d\n", flush);
	if (au->callback) {
                while (auparse_next_event(au) > 0) {
		     (*au->callback)(au, AUPARSE_CB_EVENT_READY, au->callback_user_data);
	        }
	}

     ...
}
```

This way the usercode has the freedom to choose a non-callback based way of using `auparse_feed` method.

**Comment by neel** [comment]
> However, using AUSOURCE_BUFFER can be problematic if you do not contain whole events in them.

Yes that is the problem I am encountering too. I haven't started fixing that yet as I was working on something else. In my case I have a collector that collects events and transmits over ZMQ. I don't want the collector to be heavy. So, I intend to collect enough to describe an events comprehensively and then transmit. Is it enough to check the `items` key and count the number of following records and transmit once all `items` records have been collected ? I mean how can I know that given one record sufficient information about the record (e.g. the arguments and the returns of the syscall) is captutured ? 

I am mainly interested in syscalls.

**Comment by stevegrubb** [comment]
The problem of event completeness is really what auparse was designed to do. You shouldn't need to know all the details of how the kernel creates events to start a project. Auparse insulates you from these details.

If the various threads are all from different data sources, you can instantiate multiple auparse_state_t variables by calling auparse_init for each data source. The problem would be that at some point the events need to be serialized with each other.

Also, I would be open to ideas on how serialize multiple threads at the auparse_feed API. Perhaps the callback could choose from a pool of threads?

**Comment by neel** [comment]
> Also, I would be open to ideas on how serialize multiple threads at the auparse_feed API. Perhaps the callback could choose from a pool of threads?

I think a resources pool would be better than a thread pool. Parsing the feed requires state that cannot be shared by two threads simultaneously. So, a pool of auparse states would be better I guess. The entry to these states will be guarded by a counting semaphore. Usercode may have a threadpool from where they invoke auparse_feed function. That thread spawned by the usercode will either wait for semaphore or will process it on the available auparse state. I've recently designed a similar resource pool for Lua. But it is a bit more complicated due to type bindings and its in C++.

My design is actually like this. I am collecting audit logs through an agent. I want to minimize the complexity of collection, because there can be many other collectors, e.g. log files, pcap etc.. Rather I want to collect and transmit with minimal processing. If I use auparse, while collecting then I have to serialize it and then transmit to the server. But the server will deserialize it anyway and the audit log is already serialized and parsable. So, I am avoiding parsing it while collecting. Rather I'd parse it on the server.

I have one question.

I am only interested in syscall. Is it okay if the collector first checks whether the entry is of type SYSCALL or not. If it is then it creates a bucket in the dictionary. After that, if it gets an entry having the same msg id then it puts it under the same bucket. It closes the bucket once it gets a PROCTITLE type entry. Once the bucket is closed take all the entries and send it over zmq. Once I receive it on the server I process it using auparse. There I use `AUSOURCE_BUFFER`. So, I expect that all entries that I'll process in the server after one receive operation are associated with a single syscall event.

Is it possible to get any information about the event after the PROCTITLE has been written ? Then should I wait for few seconds more before closing the bucket ?

**Comment by stevegrubb** [comment]
@neel I am working on the auparse multi-threading problem. The patch is bigger than I suspected. It touches lots of places. So, the plan is to do a release next week around May 28 to clear out things that are ready to go. Then we can start landing several patches that are going to need to spend some time being checked out by the community before we do a release. You can help test it when it lands.

**Comment by stevegrubb** [comment]
@mxlgv That is a different problem and might need it's own issue to sort out.

**Comment by stevegrubb** [comment]
@mxlgv I know it has been a while on this, but commit 2bb89007 should address the problem you reported. It will be in the 4.1.5 release. Still mulling options for the original reported issue.
